fe/src/main/java/org/apache/doris/alter/MaterializedViewHandler.java

```diff
@@ -590,15 +767,130 @@ protected void runAfterCatalogReady() {
+     *  false, parent job is cancelled or current job hasn't finished
+     */
+    private boolean checkRollupJobDependency(RollupJobV2 rollupJobV2) {
+        RollupJobV2 parentJobV2 = (RollupJobV2) alterJobsV2.get(rollupJobV2.getParentRollupJobId());
+        if (parentJobV2 != null) {
+            // check whether current rollup job's parent rollup job is finished
+            if (parentJobV2.getJobState() == AlterJobV2.JobState.CANCELLED) {
+                rollupJobV2.cancel(String.format("rollup %s 's base rollup %s is cancelled", rollupJobV2.getRollupIndexId(), rollupJobV2.getBaseIndexId()));
+                return false;
+            }
+            if (parentJobV2.getJobState() != AlterJobV2.JobState.FINISHED) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    private void runAlterJobWithConcurrencyLimit(RollupJobV2 rollupJobV2) {
+        Set<Long> tableRunningJobSet = tableRunningJobMap.get(rollupJobV2.getTableId());
+        if (tableRunningJobSet == null) {
+            tableRunningJobSet = new HashSet<>();
+            tableRunningJobMap.put(rollupJobV2.getTableId(), tableRunningJobSet);
+        }
+
+        // current job is already in running
+        if (tableRunningJobSet.contains(rollupJobV2.getJobId())) {
+            rollupJobV2.run();
+        } else if (tableRunningJobSet.size() < Config.max_running_rollup_job_num) {
+            // create rollup index for job in case 1,see processBatchAddRollup
+            if (rollupJobV2.getPartitionIdToRollupIndex().size() == 0) {
+                Catalog catalog = Catalog.getCurrentCatalog();
+                Database db = Catalog.getInstance().getDb(rollupJobV2.dbId);
+                OlapTable olapTable = (OlapTable) db.getTable(rollupJobV2.getTableId());
+                if (!createMaterializedIndexForRollupJobV2(olapTable, db, catalog, rollupJobV2)) {
+                    rollupJobV2.cancel(String.format("rollup %s 's base rollup %s is cancelled due to create mv index failed, jobId = %s",
+                            rollupJobV2.getRollupIndexId(), rollupJobV2.getBaseIndexId(), rollupJobV2.getJobId()));
+                }
+                // for job in case 2 in log replay,they are in pending but not create index, can't transfer to WAITING_TXN,so first need transfer to pending
+                rollupJobV2.setJobState(AlterJobV2.JobState.PENDING);
+                catalog.getEditLog().logAlterJob(rollupJobV2);
+            }
+            // add current job to running queue
+            tableRunningJobSet.add(rollupJobV2.getJobId());
+            rollupJobV2.run();
+        }
+    }
+
     private void runAlterJobV2() {
-        Iterator<Map.Entry<Long, AlterJobV2>> iter = alterJobsV2.entrySet().iterator();
-        while (iter.hasNext()) {
-            Map.Entry<Long, AlterJobV2> entry = iter.next();
-            AlterJobV2 alterJob = entry.getValue();
+        replayJobDependencyAfterCatalogReady();
```

Comment:
why replay here？

Reply:
In this implementation,when fe restart,the job dependncy is lost,need to rebuild before running, set every job's parent job;
But if use "whether base rollup can be found in catalog" to judge whether parent rollup is ready is ok,so this method can be removed.